crates/ibc/src/msg.rs

```diff
@@ -91,103 +90,86 @@ impl BorshDeserialize for MsgNftTransfer {
-    pub transfer: Option<ShieldingTransfer>,
-}
-
-impl BorshSerialize for MsgAcknowledgement {
-    fn serialize<W: std::io::Write>(
-        &self,
-        writer: &mut W,
-    ) -> std::io::Result<()> {
-        let encoded_msg = self.message.clone().encode_vec();
-        let members = (encoded_msg, self.transfer.clone());
-        BorshSerialize::serialize(&members, writer)
+impl From<IbcShieldingData> for String {
+    fn from(data: IbcShieldingData) -> Self {
+        HEXUPPER.encode(&data.serialize_to_vec())
     }
 }
 
-impl BorshDeserialize for MsgAcknowledgement {
-    fn deserialize_reader<R: std::io::Read>(
-        reader: &mut R,
-    ) -> std::io::Result<Self> {
-        use std::io::{Error, ErrorKind};
-        let (msg, transfer): (Vec<u8>, Option<ShieldingTransfer>) =
-            BorshDeserialize::deserialize_reader(reader)?;
-        let message = IbcMsgAcknowledgement::decode_vec(&msg)
-            .map_err(|err| Error::new(ErrorKind::InvalidData, err))?;
-        Ok(Self { message, transfer })
+/// Extract MASP transaction from IBC envelope
+pub fn extract_masp_tx_from_envelope(
+    envelope: &MsgEnvelope,
+) -> Option<MaspTransaction> {
+    match envelope {
+        MsgEnvelope::Packet(packet_msg) => match packet_msg {
+            PacketMsg::Recv(msg) => {
+                extract_masp_tx_from_packet(&msg.packet, false)
+            }
+            PacketMsg::Ack(msg) => {
+                extract_masp_tx_from_packet(&msg.packet, true)
+            }
+            PacketMsg::Timeout(msg) => {
+                extract_masp_tx_from_packet(&msg.packet, true)
+            }
+            _ => None,
+        },
+        _ => None,
     }
 }
 
-/// IBC timeout packet message with `Transfer` for refunding to a shielded
-/// address
-#[derive(Debug, Clone)]
-pub struct MsgTimeout {
-    /// IBC timeout message
-    pub message: IbcMsgTimeout,
-    /// Shieleded transfer for MASP transaction
-    pub transfer: Option<ShieldingTransfer>,
-}
-
-impl BorshSerialize for MsgTimeout {
-    fn serialize<W: std::io::Write>(
-        &self,
-        writer: &mut W,
-    ) -> std::io::Result<()> {
-        let encoded_msg = self.message.clone().encode_vec();
-        let members = (encoded_msg, self.transfer.clone());
-        BorshSerialize::serialize(&members, writer)
+/// Decode IBC shielding data from the memo string
+pub fn decode_masp_tx_from_memo(
+    memo: impl AsRef<str>,
+) -> Option<IbcShieldingData> {
+    let bytes = HEXUPPER.decode(memo.as_ref().as_bytes()).ok()?;
+    IbcShieldingData::try_from_slice(&bytes).ok()
+}
+
+/// Extract MASP transaction from IBC packet memo
+pub fn extract_masp_tx_from_packet(
+    packet: &Packet,
+    is_sender: bool,
+) -> Option<MaspTransaction> {
+    let is_ft_packet = if is_sender {
+        packet.port_id_on_a == PortId::transfer()
+    } else {
+        packet.port_id_on_b == PortId::transfer()
+    };
+
+    let shielding_data = if is_ft_packet {
```

Comment:
There could be more packet types in the future; this logic is not very resilient.

Reply:
We should also separate this out into a function.

Reply:
Fixed: ca795542d1b2dfe1b4d3b4fea8f3dfd751672db1